connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDeserializer.scala

```diff
@@ -0,0 +1,357 @@
+import com.google.protobuf.Descriptors._
+import com.google.protobuf.Descriptors.FieldDescriptor.JavaType._
+
+import org.apache.spark.sql.catalyst.{InternalRow, NoopFilters, StructFilters}
+import org.apache.spark.sql.catalyst.expressions.{SpecificInternalRow, UnsafeArrayData}
+import org.apache.spark.sql.catalyst.util.{ArrayBasedMapData, ArrayData, GenericArrayData}
+import org.apache.spark.sql.protobuf.utils.ProtobufUtils
+import org.apache.spark.sql.protobuf.utils.ProtobufUtils.ProtoMatchedField
+import org.apache.spark.sql.protobuf.utils.ProtobufUtils.toFieldStr
+import org.apache.spark.sql.protobuf.utils.SchemaConverters.IncompatibleSchemaException
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+private[sql] class ProtobufDeserializer(
+    rootDescriptor: Descriptor,
+    rootCatalystType: DataType,
+    filters: StructFilters) {
+
+  def this(rootDescriptor: Descriptor, rootCatalystType: DataType) = {
+    this(rootDescriptor, rootCatalystType, new NoopFilters)
+  }
+
+  private val converter: Any => Option[InternalRow] =
+    try {
+      rootCatalystType match {
+        // A shortcut for empty schema.
+        case st: StructType if st.isEmpty =>
+          (_: Any) => Some(InternalRow.empty)
+
+        case st: StructType =>
+          val resultRow = new SpecificInternalRow(st.map(_.dataType))
+          val fieldUpdater = new RowUpdater(resultRow)
+          val applyFilters = filters.skipRow(resultRow, _)
+          val writer = getRecordWriter(rootDescriptor, st, Nil, Nil, applyFilters)
+          (data: Any) => {
+            val record = data.asInstanceOf[DynamicMessage]
+            val skipRow = writer(fieldUpdater, record)
+            if (skipRow) None else Some(resultRow)
+          }
+      }
+    } catch {
+      case ise: IncompatibleSchemaException =>
+        throw new IncompatibleSchemaException(
+          s"Cannot convert Protobuf type ${rootDescriptor.getName} " +
+            s"to SQL type ${rootCatalystType.sql}.",
+          ise)
+    }
+
+  def deserialize(data: Message): Option[InternalRow] = converter(data)
+
+  private def newArrayWriter(
+      protoField: FieldDescriptor,
+      protoPath: Seq[String],
+      catalystPath: Seq[String],
+      elementType: DataType,
+      containsNull: Boolean): (CatalystDataUpdater, Int, Any) => Unit = {
+
+    val protoElementPath = protoPath :+ "element"
+    val elementWriter =
+      newWriter(protoField, elementType, protoElementPath, catalystPath :+ "element")
+    (updater, ordinal, value) =>
+      val collection = value.asInstanceOf[java.util.Collection[Any]]
+      val result = createArrayData(elementType, collection.size())
+      val elementUpdater = new ArrayDataUpdater(result)
+
+      var i = 0
+      val iterator = collection.iterator()
+      while (iterator.hasNext) {
+        val element = iterator.next()
+        if (element == null) {
+          if (!containsNull) {
+            throw new RuntimeException(
```

Comment:
@SandishKumarHN Ideally we should use the latest error message framework with error classes: https://github.com/apache/spark/blob/master/core/src/main/resources/error/error-classes.json

Reply:
@gengliangwang made changes

Reply:
For all the exceptions(analysis/runtime/etc...), we need to move to error class..
https://github.com/apache/spark/pull/37840
https://github.com/apache/spark/pull/37887
But this PR is already big. Feel feel to do it in follow-up PR

Reply:
Thanks @gengliangwang. @SandishKumarHN we can do this as a follow up. You can decide. Please add a TODO here so that we get back.

Reply:
@rangadi @gengliangwang yes, will make this change in the follow-up PRs.